New calling convention for Python dispatcher

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85133
* #85132

Instead of calling into the Python dispatcher for EVERY dispatcher
call, we now have a two step process.  First, we
getattr(op: OpOverload, dispatch_key) to "load" the handler for the
function.  This can either be a conventional function (in which
case we will call it, in the same way the old Python dispatcher
worked), or it can be a DispatchKey, in which case we will directly
call that DispatchKey in C++, bypassing marshalling between Python
and C++ entirely.  OpOverload.__getattr__ is carefully written so
that it will cache the

A further optimization would be to define __slots__ on OpOverload,
and ensuring that the DispatchKey strings are interned.

The resulting Python dispatcher is less flexible: after the first
lookup, the handler is cached and we won't recompute it.  Furthermore,
by default, dispatches will not go into Python, and so you won't
get stack frames for the Python dispatcher by default.  But we get
a huge performance improvement: on the following microbenchmark
we go from 2.5s to 1.9s.

```
import time
import torch
from functorch import make_fx

def f(x):
    for i in range(1000):
        x = x * x
    return x

begin = time.time()
res = make_fx(f, tracing_mode="symbolic")(torch.randn(10, 20))
print(time.time()-begin)
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
New calling convention for Python dispatcher

Instead of calling into the Python dispatcher for EVERY dispatcher
call, we now have a two step process.  First, we
getattr(op: OpOverload, dispatch_key) to "load" the handler for the
function.  This can either be a conventional function (in which
case we will call it, in the same way the old Python dispatcher
worked), or it can be a DispatchKey, in which case we will directly
call that DispatchKey in C++, bypassing marshalling between Python
and C++ entirely.  OpOverload.__getattr__ is carefully written so
that it will cache the

A further optimization would be to define __slots__ on OpOverload,
and ensuring that the DispatchKey strings are interned.

The resulting Python dispatcher is less flexible: after the first
lookup, the handler is cached and we won't recompute it.  Furthermore,
by default, dispatches will not go into Python, and so you won't
get stack frames for the Python dispatcher by default.  But we get
a huge performance improvement: on the following microbenchmark
we go from 2.5s to 1.9s.

```
import time
import torch
from functorch import make_fx

def f(x):
    for i in range(1000):
        x = x * x
    return x

begin = time.time()
res = make_fx(f, tracing_mode="symbolic")(torch.randn(10, 20))
print(time.time()-begin)
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "New calling convention for Python dispatcher"


Instead of calling into the Python dispatcher for EVERY dispatcher
call, we now have a two step process.  First, we
getattr(op: OpOverload, dispatch_key) to "load" the handler for the
function.  This can either be a conventional function (in which
case we will call it, in the same way the old Python dispatcher
worked), or it can be a DispatchKey, in which case we will directly
call that DispatchKey in C++, bypassing marshalling between Python
and C++ entirely.  OpOverload.__getattr__ is carefully written so
that it will cache the

A further optimization would be to define __slots__ on OpOverload,
and ensuring that the DispatchKey strings are interned.

The resulting Python dispatcher is less flexible: after the first
lookup, the handler is cached and we won't recompute it.  Furthermore,
by default, dispatches will not go into Python, and so you won't
get stack frames for the Python dispatcher by default.  But we get
a huge performance improvement: on the following microbenchmark
we go from 2.5s to 1.9s.

```
import time
import torch
from functorch import make_fx

def f(x):
    for i in range(1000):
        x = x * x
    return x

begin = time.time()
res = make_fx(f, tracing_mode="symbolic")(torch.randn(10, 20))
print(time.time()-begin)
```

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "New calling convention for Python dispatcher"


Instead of calling into the Python dispatcher for EVERY dispatcher
call, we now have a two step process.  First, we
getattr(op: OpOverload, dispatch_key) to "load" the handler for the
function.  This can either be a conventional function (in which
case we will call it, in the same way the old Python dispatcher
worked), or it can be a DispatchKey, in which case we will directly
call that DispatchKey in C++, bypassing marshalling between Python
and C++ entirely.  OpOverload.__getattr__ is carefully written so
that it will cache the

A further optimization would be to define __slots__ on OpOverload,
and ensuring that the DispatchKey strings are interned.

The resulting Python dispatcher is less flexible: after the first
lookup, the handler is cached and we won't recompute it.  Furthermore,
by default, dispatches will not go into Python, and so you won't
get stack frames for the Python dispatcher by default.  But we get
a huge performance improvement: on the following microbenchmark
we go from 2.5s to 1.9s.

```
import time
import torch
from functorch import make_fx

def f(x):
    for i in range(1000):
        x = x * x
    return x

begin = time.time()
res = make_fx(f, tracing_mode="symbolic")(torch.randn(10, 20))
print(time.time()-begin)
```

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

wconstab commented on torch/__init__.py:
```diff
@@ -982,4 +982,4 @@ def _register_device_module(device_type, module):
     csan.enable_cuda_sanitizer()
 
-from ._dispatch import python
+torch._C._set_python_dispatcher(True)
```
woah this got real fast

ezyang replied:
this doesn't actually enable the dispatcher lol

wconstab commented on torch/_ops.py:
```diff
@@ -40,8 +47,60 @@ def py_impl(self, dispatch_key, fn):
+        elif not has_backend_kernel:
+            return cand
+    # 2.4. For autograd backend keys, use kernel from DispatchKey::Autograd if available
+    cand = DispatchKey.Autograd
```
it is sort of surprising that we'd use a kernel registered to the 'Autograd' key for an op, although i'm not sure what this key really means.

wconstab commented on torch/_ops.py:
```diff
@@ -198,12 +259,10 @@ def __str__(self):
-        if (
-            torch._C._dispatch_has_kernel_for_dispatch_key(self.name(), dk)
-            or dk in self.py_kernels
-        ):
```
what happened to py_kernels?

ezyang replied:
I guess I can restore consulting it

wconstab replied:
which kernels is self.py_kernels anyway, is that where non-dispatcher metas go or something?  i wasn't sure why you removed it but also not sure what it's for.

wconstab commented on torch/_ops.py:
```diff
@@ -228,22 +287,44 @@ def inner(fn):
+                return self.python_key_mode_table[curr_mode](*args, **kwargs)
+
+            setattr(self, attr, handler)
+            return handler
```
what's the deal with returning a handler function instead of a DispatchKey object?  i'm trying to find where this getattr is used and how it is determined whether to call its return value

ezyang replied:
If you return a DispatchKey object, we directly call the c++ kernel for the dispatch key. If you return a function, we call the function to handle the kernel. The getattr is triggered by https://github.com/pytorch/pytorch/pull/85133/files/d5013fcde2ff0a9d9587614f77da73b4b934d331#diff-0fa862a86ab690b67dbc2aecb5cc0675d6df08af87e8a2b0655e554cef6f11d7

bdhirsh replied:
could you add that in a comment somewhere in `def __getattr__`? (that the contract is we return either a python handler fn, or a dispatch key). The caller-callee stuff is very non-local :)

wconstab commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -2279,39 +2279,35 @@ void ConcretePyInterpreterVTable::python_dispatcher(
+  py::handle torch_api_function_overload = getTorchApiFunction(op);
+
+  c10::DispatchKey k = ks.highestPriorityTypeId();
+  auto handler = torch_api_function_overload.attr(toString(k));
```
is handler maybe a function? does it have a .ptr attribute? i'm confused here..

ezyang replied:
handler is a pybind11 `py::handle`, aka a PyObject.

wconstab replied:
ok.  i guess as long as we don't expect that getattr to get invoked from multiple call sites, i'm ok with this.

wconstab commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -2279,39 +2279,35 @@ void ConcretePyInterpreterVTable::python_dispatcher(
-      c10::impl::PythonDispatcherTLS::get_state().ptr(getPyInterpreter());
-  TORCH_INTERNAL_ASSERT(python_dispatcher);
+  py::object obj = py::reinterpret_steal<py::object>(
+      PyObject_Call(handler.ptr(), args.ptr(), kwargs.ptr()));
```
ahh. ok, i missed this part.

ezyang committed:
Update on "New calling convention for Python dispatcher"


Instead of calling into the Python dispatcher for EVERY dispatcher
call, we now have a two step process.  First, we
getattr(op: OpOverload, dispatch_key) to "load" the handler for the
function.  This can either be a conventional function (in which
case we will call it, in the same way the old Python dispatcher
worked), or it can be a DispatchKey, in which case we will directly
call that DispatchKey in C++, bypassing marshalling between Python
and C++ entirely.  OpOverload.__getattr__ is carefully written so
that it will cache the

A further optimization would be to define __slots__ on OpOverload,
and ensuring that the DispatchKey strings are interned.

The resulting Python dispatcher is less flexible: after the first
lookup, the handler is cached and we won't recompute it.  Furthermore,
by default, dispatches will not go into Python, and so you won't
get stack frames for the Python dispatcher by default.  But we get
a huge performance improvement: on the following microbenchmark
we go from 2.5s to 1.9s.

```
import time
import torch
from functorch import make_fx

def f(x):
    for i in range(1000):
        x = x * x
    return x

begin = time.time()
res = make_fx(f, tracing_mode="symbolic")(torch.randn(10, 20))
print(time.time()-begin)
```

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

bdhirsh commented on torch/_ops.py:
```diff
@@ -40,8 +47,60 @@ def py_impl(self, dispatch_key, fn):
+DispatchKey = torch._C.DispatchKey
+
+# Equivalent to computeDispatchTableEntryWithDebug
+def resolve_key(op: PyOperatorABC, k: DispatchKey):  # type: ignore[valid-type]
```
This probably isn't going to break anything in the near future, but we'll have to remember to keep these in sync:

(1) `[CompositeImplicitAutogradNestedTensor](https://github.com/pytorch/pytorch/blob/f37069aac7a0c4fd1c3455e4e9058bf56fb759f4/aten/src/ATen/core/dispatch/OperatorEntry.cpp#L330)` is missing

(2) Note to myself to add `Autocast` here if we end up making autocast an [alias key](https://github.com/pytorch/pytorch/pull/85043/files#diff-da6c1aa633224c76f5ae6b4e80676dadbb24bac021770ee01d5560d2db8ffb92R88)

ezyang replied:
It's here? 

```
    # 2.3. Use CompositeImplicitAutograd kernel if available
    cand = DispatchKey.CompositeImplicitAutogradNestedTensor
    if (
        (k != DispatchKey.Undefined and is_included_in_alias(k, cand))
        and has_key(op, cand)
        and not has_backend_kernel
    ):
        return cand
    cand = DispatchKey.CompositeImplicitAutograd
    if (k == DispatchKey.Undefined or is_included_in_alias(k, cand)) and has_key(
        op, cand
```

bdhirsh replied:
oh whoops haha

bdhirsh commented on torch/_ops.py:
```diff
@@ -218,32 +282,54 @@ def inner(fn):
+                if curr_mode not in self.python_key_mode_table:
+                    # TODO: This path is slow, should generally encourage this
+                    # case to not happen
+                    return self._op_dk(key, *args, **kwargs)
```
hmm still reading, but:

Wouldn't it be faster (and remove a case) to folder this case into here?
```
            if not self.python_key_mode_table:
                setattr(self, attr, key)
                return key
```

If I'm following correctly, it looks like when the given mode isn't registered to the PyDispatcher, we're returning a python handler that just redispatches to the python key. But it seems like the point of this PR is that we want to fast-path the case where we're just using the ordinary dispatcher, so we may as well return a dispatch key instead of a python handler when `not self.python_key_mode_table or curr_mode not in self.python_key_mode_table`.

ezyang replied:
I'm not sure exactly what you are suggesting. But if you are proposing:

```
        if key == torch._C.DispatchKey.Python:
            curr_mode = type(torch._C._get_torch_dispatch_mode()) 
            if not self.python_key_mode_table or curr_mode not in self.python_key_mode_table:
                setattr(self, attr, key)
                return key
```

This will not work. In general, the current mode may vary on subsequent invocations, so you cannot pre-compute it at the first time the getattr is called

bdhirsh replied:
ohh. That makes sense - thanks for the clarification.

ezyang committed:
Update on "New calling convention for Python dispatcher"


Instead of calling into the Python dispatcher for EVERY dispatcher
call, we now have a two step process.  First, we
getattr(op: OpOverload, dispatch_key) to "load" the handler for the
function.  This can either be a conventional function (in which
case we will call it, in the same way the old Python dispatcher
worked), or it can be a DispatchKey, in which case we will directly
call that DispatchKey in C++, bypassing marshalling between Python
and C++ entirely.  OpOverload.__getattr__ is carefully written so
that it will cache the

A further optimization would be to define __slots__ on OpOverload,
and ensuring that the DispatchKey strings are interned.

The resulting Python dispatcher is less flexible: after the first
lookup, the handler is cached and we won't recompute it.  Furthermore,
by default, dispatches will not go into Python, and so you won't
get stack frames for the Python dispatcher by default.  But we get
a huge performance improvement: on the following microbenchmark
we go from 2.5s to 1.9s.

```
import time
import torch
from functorch import make_fx

def f(x):
    for i in range(1000):
        x = x * x
    return x

begin = time.time()
res = make_fx(f, tracing_mode="symbolic")(torch.randn(10, 20))
print(time.time()-begin)
```

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
